sql/cagg_api.sql

```diff
@@ -0,0 +1,192 @@
+            _timescaledb_functions.to_unix_microseconds(upper(aligned)));
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+-- Get raw ranges from the materialization invalidation log
+--
+-- This is a cleaned-up version of the timestamps, still in Unix
+-- microseconds, with nulls for '-infinity' and '+infinity' and
+-- invalid entries removed.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_raw_materialization_ranges(typ regtype)
+RETURNS TABLE (materialization_id integer,
+               lowest_modified_value bigint,
+               greatest_modified_value bigint)
+AS $$
+   SELECT materialization_id,
+          NULLIF(lowest_modified_value, _timescaledb_functions.get_min_for_type(typ)),
+          NULLIF(greatest_modified_value, _timescaledb_functions.get_max_for_type(typ))
+     FROM _timescaledb_catalog.continuous_aggs_materialization_invalidation_log
+    WHERE lowest_modified_value
+          BETWEEN _timescaledb_functions.get_min_for_type(typ)
+              AND _timescaledb_functions.get_max_for_type(typ)
+      AND greatest_modified_value
+          BETWEEN _timescaledb_functions.get_min_for_type(typ)
+              AND _timescaledb_functions.get_max_for_type(typ)
+$$
+LANGUAGE SQL
+SET search_path TO pg_catalog, pg_temp;
+
+-- Get materialization invalidations for a continuous aggregate.
+--
+-- Note that this will modify the materialization invalidation table
+-- to be able to extract the restricted range of invalidations.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_materialization_invalidations(
+    continuous_aggregate REGCLASS,
+    restriction TSTZRANGE,
+    OUT token TEXT,
+    OUT invalidations TSTZMULTIRANGE
+) RETURNS RECORD AS
+$body$
+DECLARE
+    info RECORD := _timescaledb_functions.get_materialization_info(continuous_aggregate);
+    aligned TSTZRANGE := _timescaledb_functions.align_to_bucket(info.bucket_width, restriction);
+BEGIN
+    -- Compute the multirange for the invalidations inside the
+    -- restriction passed down to the function and store this in the
+    -- result record.
+    WITH
+      ranges AS (
+          SELECT materialization_id,
+                 range_agg(_timescaledb_functions.as_multirange(lowest_modified_value,
+                                                                greatest_modified_value,
+                                                                null::tstzrange)) AS invals
+            FROM _timescaledb_functions.get_raw_materialization_ranges('timestamptz'::regtype)
+          GROUP BY materialization_id
+      )
+    SELECT range_agg(invals * multirange(aligned)) INTO invalidations
+      FROM ranges
+     WHERE invals && aligned
+       AND materialization_id = info.materialization_id;
+
+    token := pg_current_snapshot()::text;
+END
+$body$
+LANGUAGE plpgsql
+SET search_path TO pg_catalog, pg_temp;
+
+CREATE OR REPLACE PROCEDURE _timescaledb_functions.accept_materialization_invalidations(
```

Comment:
IMO, it is not clear from the function name what this function does and what the implications are. There's also no function description. "Accept" could mean a lot of things, and there are hidden side effects like deletions happening.

It seems that the function is just deleting all invalidations visible in the given snapshot (via the token parameter), so maybe we should be explict: `delete_materialization_invalidations_in_snapshot()`?

Reply:
Also, not sure I understand why this API function needs to be passed a token/snapshot?

It almost seems like there's a requirement to first "get" invalidations (i.e., just read them). But why can't I just delete invalidations in a range at the current instant/snapshot without first reading them?

If I want to read and "accept" invalidations in a snapshot I could just simply start a transaction in REPEATABLE READ and get and accept within that transaction block.

The thing is, if I implement a refresh procedure, I would like _all_ invalidation in the refreshed range to be deleted at the time instant of the refresh. I don't want to leave some invalidations that was added between getting the token and "accepting" with the token. Then the cagg will still be invalid after the refresh.

Reply:
As discussed, we remove this function and instead rely on refresh function to update the invalidations.